src/vsr/replica_test.zig

```diff
@@ -1355,6 +1355,68 @@ test "Cluster: upgrade: state-sync to new release" {
     try expectEqual(t.replica(.R2).commit(), t.replica(.R_).commit());
 }
 
+test "Cluster: scrub: background scrubber, fully corrupt grid" {
+    const t = try TestContext.init(.{ .replica_count = 3 });
+    defer t.deinit();
+
+    var c = t.clients(0, t.cluster.clients.len);
+    try c.request(checkpoint_2_trigger, checkpoint_2_trigger);
+    try expectEqual(t.replica(.R_).commit(), checkpoint_2_trigger);
+
+    var a0 = t.replica(.A0);
+    var b2 = t.replica(.B2);
+
+    const a0_free_set = &t.cluster.replicas[a0.replicas.get(0)].grid.free_set;
+    const b2_free_set = &t.cluster.replicas[b2.replicas.get(0)].grid.free_set;
+    const b2_storage = &t.cluster.storages[b2.replicas.get(0)];
+
+    // Corrupt B2's entire grid.
+    // Note that we intentionally do *not* shut down B2 for this – the intent is to test the
+    // scrubber, without leaning on Grid.read_block()'s `from_local_or_global_storage`.
+    {
+        var address: u64 = 1;
+        while (address <= Storage.grid_blocks_max) : (address += 1) {
+            b2.corrupt(.{ .grid_block = address });
+        }
+    }
+
+    // Disable new read/write faults so that we can use `storage.faults` to track repairs.
+    // (That is, as the scrubber runs, the number of faults will monotonically decrease.)
+    b2_storage.options.read_fault_probability = 0;
+    b2_storage.options.write_fault_probability = 0;
+
+    // Tick until B2's grid repair stops making progress.
+    // (This may take multiple scrubber cycles, since when a table index is corrupt, we skip over
+    // scrubbing its data for that cycle.)
+    {
+        var faults_before = b2_storage.faults.count();
+        while (true) {
+            t.run();
+
+            var faults_after = b2_storage.faults.count();
+            assert(faults_after <= faults_before);
+            if (faults_after == faults_before) break;
+
+            faults_before = faults_after;
+        }
+    }
+
+    // Verify that B2 repaired all blocks.
+    var address: u64 = 1;
+    while (address <= Storage.grid_blocks_max) : (address += 1) {
+        if (a0_free_set.is_free(address)) {
+            assert(b2_free_set.is_free(address));
+            assert(b2_storage.area_faulty(.{ .grid = .{ .address = address } }));
+        } else {
+            assert(!b2_free_set.is_free(address));
+            assert(!b2_storage.area_faulty(.{ .grid = .{ .address = address } }));
+        }
+    }
+
+    try TestReplicas.expect_equal_grid(a0, b2);
+    try TestReplicas.expect_equal_grid(b2, b2);
```

Comment:
Is comparing b2 with itself intentional?

Reply:
Nope; typo.